### PR TITLE
Improve ABI types in wasmv1

### PR DESCRIPTION
**Pull request by bilboquet**

### CODE REVIEW

File: proto/abis/massa/abi/v1/abi.proto [hunk 1]
```diff
@@ -82,22 +74,257 @@ message GenerateEventRequest {
...
+  string function_name = 2;
 }
 
 // FunctionExistsResponse
 message FunctionExistsResponse {
   // Exists is true if the function exists
   bool exists = 1;
 }
+
+// Call RespResult becaule Result will conflict with the Result type in Rust
```
Comment (aoudiamoncef) [inline proto/abis/massa/abi/v1/abi.proto:94]: Because

**Review by aoudiamoncef — APPROVED** [review]
LGTM good job 👏 

Just add missing comments on messages and field and It's ready to ship

**Comment by damip** [comment]
```
* use https://protobuf.dev/reference/csharp/api-docs/class/google/protobuf/well-known-types/empty instead of our own Empty

* use bytes in GenerateEvent

* SubNativeAmountsRequest
	use "left" and "right" and explain that we compute "left - right, fail if right > left" instead of "minuend/subtrahend"
 

* NativeAmountToBytesRequest
	my bad; technically the SC can already do byte conversion !

* remove LogRequest: we are deprecating log printing in favor of Events
```

**Comment by bilboquet** [comment]
> LGTM good job clap
> 
> Just add missing comments on messages and field and It's ready to ship

Yes, it's still draft, because up to now I was still concerned by the structure of the .proto. (it was hardly extensible and not that easy to manipulate in Rust)
I think it's now ok, hence the PR :)
And so I can:
[] add comments
[] verify int vs fixed

File: proto/abis/massa/abi/v1/abi.proto (hunk 1 above)
Comment (bilboquet) [reply proto/abis/massa/abi/v1/abi.proto:94]: x) 
(probably copilot fault ;)

File: proto/abis/massa/abi/v1/abi.proto [hunk 2]
```diff
@@ -82,22 +74,257 @@ message GenerateEventRequest {
 
 // TransferCoins
```
[the remark this answers is not in the document]
Comment (bilboquet) [reply proto/abis/massa/abi/v1/abi.proto:75]: Those first requests are already fully implemented in the wasmv1. (we can write sc with them).
As testing is hard, I will first use them to tests amount, address… then I'll adjust them.
But yes most of them must be adapted so I keep this comment open ^^.

**Comment by aoudiamoncef** [comment]
> 

IMHO, using well known type `Empty` could be a breaking change in the future. 

In the first version private API, I tried to use a xxxRequest xxxResponse even it's an empty. If we would like to change input/output of the method = no breaking change

File: proto/abis/massa/abi/v1/abi.proto [hunk 3]
```diff
@@ -13,65 +19,51 @@ option php_namespace = "Com\\Massa\\Abi\\V1";
...
-  string address = 1;
-}
-
-// Amount
-message Amount {
-  // Amount is a 64-bit unsigned integer
-  fixed64 amount = 1;
+// Error message used in abi Response
+message Error {
+  string message = 1;
```
[the remark this answers is not in the document]
Comment (bilboquet) [reply proto/abis/massa/abi/v1/abi.proto:24]: done

**Comment by bilboquet** [comment]
> > LGTM good job clap
> > Just add missing comments on messages and field and It's ready to ship
> 
> Yes, it's still draft, because up to now I was still concerned by the structure of the .proto. (it was hardly extensible and not that easy to manipulate in Rust) I think it's now ok, hence the PR :) And so I can: 

[X] add comments 
[X] verify int vs fixed

File: proto/abis/massa/abi/v1/abi.proto [hunk 4]
```diff
@@ -13,65 +18,54 @@ option php_namespace = "Com\\Massa\\Abi\\V1";
...
 }
 
 // LocalCallResponse
 message LocalCallResponse {
   // Return_data is the return value of the function
-  bytes return_data = 1;
+  bytes data = 1;
 }
 
 // GenerateEventRequest
```
Comment (bilboquet) [inline proto/abis/massa/abi/v1/abi.proto:71]: use bytes in GenerateEvent but have the default SDK take strings and convert them to UTF8 when calling the ABI

**Comment by bilboquet** [comment]
> * use https://protobuf.dev/reference/csharp/api-docs/class/google/protobuf/well-known-types/empty instead of our own Empty 

Own type removed, will use the one from google when needed (if the discussion in the other comment is correct)
[X] done

> * use bytes in GenerateEvent but have the default SDK take strings and convert them to UTF8 when calling the ABI 

added a dedicated comment to track this point (so consider this one closed)
[X] done

> * SubNativeAmountsRequest use "left" and "right" and explain that we compute "left - right, fail if right 
> left" instead of "minuend/subtrahend" 
[X] done

> * NativeAmountToBytesRequest technically the SC can already do to/from byte conversion ! I would vote for removing those to/from byte conversions and just use the builtin SC-based deserializer together with the Check * ABIs to verify the integrity of the data 

I took those 2 from the original discussion without much thinking... I fully agree that if the SC can do it by itself we can remove them. 
[X] done

> * remove LogRequest: we are deprecating log printing in favor of Events`

moved in a separate branch as this is very useful when writing the sdk, but for sure it should not be in the final ABI

File: proto/abis/massa/abi/v1/abi.proto (hunk 4 above)
Comment (bilboquet) [reply proto/abis/massa/abi/v1/abi.proto:71]: @damip 
just for information, the current wip implementation of this ABI in the AS SDK already does the conversion from UTF16 (user side) <--> UTF8 (abi side)
basically it's just adding a `to_vec`

File: proto/abis/massa/abi/v1/abi.proto [hunk 5]
```diff
@@ -13,65 +18,54 @@ option php_namespace = "Com\\Massa\\Abi\\V1";
...
+message Error {
+  // a string representing the error
+  string message = 1;
 }
 
-// Empty
-message Empty {}
-
-// CreateSC
 message CreateSCRequest {
```
Comment (bilboquet) [inline proto/abis/massa/abi/v1/abi.proto:27]: Transform to the Response<Result,Error> form (later)

File: proto/abis/massa/abi/v1/abi.proto [hunk 6]
```diff
@@ -13,65 +18,54 @@ option php_namespace = "Com\\Massa\\Abi\\V1";
...
-// CreateSCResponse
 message CreateSCResponse {
-  // Address is a string representation of the address
-  Address address = 1;
+  // Address of the just created smart contract
+  massa.model.v1.NativeAddress sc_address = 1;
 }
 
 // CallSC
 message CallRequest {
```
Comment (bilboquet) [inline proto/abis/massa/abi/v1/abi.proto:38]: DON'T Transform to the Response<Result,Error> form

File: proto/abis/massa/abi/v1/abi.proto [hunk 7]
```diff
@@ -13,65 +18,54 @@ option php_namespace = "Com\\Massa\\Abi\\V1";
...
 
 // CallResponse
 message CallResponse {
   // Return_data is the return value of the function
-  bytes return_data = 1;
+  bytes data = 1;
 }
 
 // LocalCall
 message LocalCallRequest {
```
Comment (bilboquet) [inline proto/abis/massa/abi/v1/abi.proto:56]: DON'T Transform to the Response<Result,Error> form

File: proto/abis/massa/abi/v1/abi.proto (hunk 4 above)
Comment (damip) [reply proto/abis/massa/abi/v1/abi.proto:71]: Basically in the SC SDK we would take a utf16 string as argument and convert it to utf8 then to bytes before sending it to the abi

Comment (bilboquet) [reply proto/abis/massa/abi/v1/abi.proto:71]: That's what I was trying to say :)

File: proto/abis/massa/abi/v1/abi.proto [hunk 8]
```diff
@@ -82,22 +76,319 @@ message GenerateEventRequest {
 
 // TransferCoins
 message TransferCoinsRequest {
```
Comment (bilboquet) [inline proto/abis/massa/abi/v1/abi.proto:78]: Transform to the Response<Result,Error> form (later)

File: proto/abis/massa/abi/v1/abi.proto [hunk 9]
```diff
@@ -392,3 +399,107 @@ message DivRemNativeAmountResult {
   // Remainder of amount and divisor
   massa.model.v1.NativeAmount remainder = 2;
 }
+
+// Comparison result
+message ComparisonResult {
```
Comment (bilboquet) [inline proto/abis/massa/abi/v1/abi.proto:404]: @aoudiamoncef 
Do you think I should put this one in commons?
I feel like it should be

Comment (damip) [reply proto/abis/massa/abi/v1/abi.proto:404]: Yes I think so too, a bit like https://doc.rust-lang.org/std/cmp/enum.Ordering.html

Comment (bilboquet) [reply proto/abis/massa/abi/v1/abi.proto:404]: let's agree on the path before I move it
I propose 
`proto/commons/massa/shared/v1/comparison.proto`

it follow the scheme of others
`proto/abis/massa/abi/v1/abi.proto`
`proto/abis/massa/abi/v1/abi.proto`
`proto/commons/massa/model/v1/address.proto`...

File: proto/abis/massa/abi/v1/abi.proto [hunk 10]
```diff
@@ -82,22 +77,519 @@ message GenerateEventRequest {
...
+// keccak256 hash request
+message Keccak256Request {
+  // Data to hash
+  bytes data = 1;
+}
+
+// keccak256 hash result
+message Keccak256Result {
+  // Hash of data
+  bytes hash = 1;
```
Comment (bilboquet) [inline proto/abis/massa/abi/v1/abi.proto:594]: @damip 
What about returning a NativeHash?

File: proto/abis/massa/abi/v1/abi.proto [hunk 11]
```diff
@@ -13,91 +19,638 @@ option php_namespace = "Com\\Massa\\Abi\\V1";
...
+}
+
+// Keccak256 hash result
+message Keccak256Result {
+  // Hash of data
+  bytes hash = 1;
+}
+
+// EVM signature verification request
+message VerifyEvmSigRequest {
```
Comment (bilboquet) [inline proto/abis/massa/abi/v1/abi.proto:611]: For every Verify*SigRequest
put Sig as the first parameter

File: proto/abis/massa/abi/v1/abi.proto [hunk 12]
```diff
@@ -13,65 +19,51 @@ option php_namespace = "Com\\Massa\\Abi\\V1";
...
 message Empty {}
 
-// CreateSC
 message CreateSCRequest {
   // Bytecode is the compiled code of the smart contract
   bytes bytecode = 1;
 }
 
-// CreateSCResponse
 message CreateSCResponse {
```
[the remark this answers is not in the document]
Comment (bilboquet) [reply proto/abis/massa/abi/v1/abi.proto:33]: done

**Pull request closed without merging**